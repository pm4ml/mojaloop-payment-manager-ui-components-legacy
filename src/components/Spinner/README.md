## Spinner

Renders a infinite spinning circle path

### Proptypes

|     Name    |        Type        |                        Description                        | Optional | Values |   Default   |
|-------------|--------------------|-----------------------------------------------------------|----------|--------|-------------|
| size        | `string`, `number` | Describes the size in pixels                              |          |        | `20`        |
| center      | `bool`             | Describes whether to render centrally in a flex container |          |        | `false`     |
| color       | `string`           | Describes the hex color                                   |          |        | `undefined` |