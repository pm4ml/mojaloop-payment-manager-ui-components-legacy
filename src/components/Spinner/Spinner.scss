@import '../../assets/styles/vars/colors.scss';

.el-spinner {
  animation: rotateSpinner 4s linear infinite;

  &.center {
    display: flex;
    width: 100%;
    height: inherit;
    align-items: center;
    justify-content: center;
  }
}
.el-spinner__component {
  animation: rotateSpinner 0.8s cubic-bezier(0.53, 0.26, 0.54, 0.74) infinite;
}

.el-spinner__svg-path {
  fill: none;
  stroke-dasharray: 100;
  stroke-dashoffset: 0;
  transform-origin: center;
  stroke-linecap: round;
  stroke: $color__main-primary;
}

@keyframes rotateSpinner {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}
