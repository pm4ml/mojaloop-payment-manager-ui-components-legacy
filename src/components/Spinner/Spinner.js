import './Spinner.scss';

import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';

const polarToCartesian = (centerX, centerY, radius, angleInDegrees) => {
  const angleInRadians = (angleInDegrees - 90) * (Math.PI / 180.0);
  return {
    x: centerX + radius * Math.cos(angleInRadians),
    y: centerY + radius * Math.sin(angleInRadians),
  };
};

const describeArc = (x, y, radius, startAngle, endAngle) => {
  const start = polarToCartesian(x, y, radius, endAngle);
  const end = polarToCartesian(x, y, radius, startAngle);
  const largeArcFlag = endAngle - startAngle <= 180 ? '0' : '1';
  return ['M', start.x, start.y, 'A', radius, radius, 0, largeArcFlag, 0, end.x, end.y].join(' ');
};

class Spinner extends PureComponent {
  render() {
    const { size, center, color } = this.props;
    const realSize = typeof size === 'string' ? { s: 20, m: 40, l: 60 }[size] : size;
    const strokeWidth = realSize / 10;
    const width = `${realSize}px`;
    const height = `${realSize}px`;
    const position = realSize / 2;
    const radius = position - strokeWidth;
    const spinnerStyle = {
      width: center === false && width,
      height: center === false && height,
    };
    const pathStyle = { stroke: color };

    return (
      <div className={`el-spinner ${center ? 'center' : ''}`} style={spinnerStyle}>
        <svg
          className="el-spinner__component"
          width={width}
          height={height}
          viewBox={`0 0 ${realSize} ${realSize}`}
        >
          <path
            className="el-spinner__svg-path"
            strokeWidth={strokeWidth}
            d={describeArc(position, position, radius, 90, 200)}
            style={pathStyle}
          />
        </svg>
      </div>
    );
  }
}

Spinner.defaultProps = {
  size: 20,
  center: false,
  color: undefined,
};

Spinner.propTypes = {
  size: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  center: PropTypes.bool,
  color: PropTypes.string,
};
export default Spinner;
