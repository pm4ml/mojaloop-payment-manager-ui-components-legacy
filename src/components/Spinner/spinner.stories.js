import React from 'react';

import Row from '../Row';
import Spinner from './Spinner';

export default {
  title: 'Spinner',
  component: Spinner,
};

export const Default = () => <Spinner />;
export const Colors = () => (
  <Row align="left">
    <Spinner color="#c33" />
    <Spinner color="#3c3" />
    <Spinner color="#33c" />
    <Spinner color="#cc3" />
    <Spinner color="#3cc" />
  </Row>
);

export const PresetSizes = () => (
  <Row align="left">
    <Spinner size="s" />
    <Spinner size="m" />
    <Spinner size="l" />
  </Row>
);

export const NumericSizes = () => (
  <Row align="left">
    <Spinner size={10} />
    <Spinner size={20} />
    <Spinner size={30} />
    <Spinner size={40} />
    <Spinner size={50} />
    <Spinner size={60} />
    <Spinner size={100} />
  </Row>
);

export const CenteredRelativeToParent = () => (
  <div
    style={{
      height: '400px',
      width: '400px',
      border: '5px dashed #ccc',
      background: 'linear-gradient(30deg, #f8f8f8, #d8d8d8)',
    }}
  >
    <Spinner size={100} center />
  </div>
);
