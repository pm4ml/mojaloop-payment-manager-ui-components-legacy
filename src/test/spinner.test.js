import { mount, shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import React from 'react';

import Spinner from '../components/Spinner';

it('renders the spinner', () => {
  const wrapper = shallow(<Spinner />);
  expect(wrapper.find('.el-spinner')).toBeTruthy();
});

it('renders the spinner elements with correct size', () => {
  [1, 2, 3, 10, 20, 30, 100, 200, 300].forEach(size => {
    const wrapper = shallow(<Spinner size={size} />);
    const svg = wrapper.find('svg');
    const style = wrapper.find('.el-spinner').prop('style');
    expect(svg.prop('height')).toBe(`${size}px`);
    expect(svg.prop('width')).toBe(`${size}px`);
    expect(style).toHaveProperty('width', `${size}px`);
    expect(style).toHaveProperty('height', `${size}px`);
  });
});

it('renders the spinner with color', () => {
  const wrapper = shallow(<Spinner color="#f00" />);
  expect(wrapper.find('path').prop('style')).toHaveProperty('stroke', '#f00');
});

it('renders the spinner centered to the parent container', () => {
  const wrapper = mount(
    <div style={{ height: '100px', width: '100px' }}>
      <Spinner size={20} center />
    </div>,
  );
  expect(wrapper.find('.center')).toHaveLength(1);
  const svg = wrapper.find('svg');
  expect(svg.prop('height')).toBe('20px');
  expect(svg.prop('width')).toBe('20px');
});

// Snapshot
it('renders the button correctly when multiple props are set', () => {
  const wrapper = shallow(<Spinner color="#ff0" size={30} />);
  expect(shallowToJson(wrapper)).toMatchSnapshot();
});
