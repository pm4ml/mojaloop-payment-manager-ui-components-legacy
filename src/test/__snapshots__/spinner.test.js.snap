// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the button correctly when multiple props are set 1`] = `
<div
  className="el-spinner "
  style={
    Object {
      "height": "30px",
      "width": "30px",
    }
  }
>
  <svg
    className="el-spinner__component"
    height="30px"
    viewBox="0 0 30 30"
    width="30px"
  >
    <path
      className="el-spinner__svg-path"
      d="M 10.895758280091975 26.2763114494309 A 12 12 0 0 0 27 15"
      strokeWidth={3}
      style={
        Object {
          "stroke": "#ff0",
        }
      }
    />
  </svg>
</div>
`;
